### PR TITLE
add custom closure compiler runner

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/2386

### CODE REVIEW

File: third_party/closure-compiler/src/AmpCommandLineRunner.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      "dev.fine",
+      "dev.assert");
+
+  protected AmpCommandLineRunner(String[] args) {
+    super(args);
+  }
+
+  @Override
+  protected CompilerOptions createOptions() {
+    CompilerOptions options = super.createOptions();
```
Comment (dvoytenko) [inline third_party/closure-compiler/src/AmpCommandLineRunner.java]: How do we test this? Should we first try it w/o LOG_STRIP_TYPES and see if the result is equivalent?

Comment (erwinmombay) [inline third_party/closure-compiler/src/AmpCommandLineRunner.java]: @dvoytenko absolutely. trying to get a local green test first before since local is super flakey right now =/. i also need to add its integration with the gulp build pipeline

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      "dev.assert");
+
+  protected AmpCommandLineRunner(String[] args) {
+    super(args);
+  }
+
+  @Override
+  protected CompilerOptions createOptions() {
+    CompilerOptions options = super.createOptions();
+    options.setStripTypes(Sets.newHashSet(LOG_STRIP_SUFFIX));
```
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: We have 2 mechanisms to do DCE, `setStripTypes/setStripTypePrefixes` (option A) and `setStripNameSuffixes/setStripNamePrefixes` (option B). option A is more safer as we can declare an expression like `dev.fine` while option b seems to only allow `fine` or `assert` (no access operation, just method name).

the issue is that the strip operation seems to occur **after** the `SIMPLE_PLUS_OPTIMIZATION` has ran, meaning if you did `setStripTypes('dev.fine')` (option A), it wouldn't work, as the actual js code it operates on would have be `a.fine` (after var renaming). Now we could use option B, but option B is extremely unsafe since `fine` or `assert` are very generic and any other module/class could declare one, and to get around this we would need the `log` module to declare method names like `devfine` and `devassert` (even then it feels extremely unsafe).

One other option we can do is assign the log instances to the global `AMP` object since that is never renamed and an expression like `AMP.dev.fine('hello world')` would be stripped out if we do `setStripTypes('AMP.dev.fine'). But this would also mean `log` has to be the very first module loaded since it would be needed by the other modules that uses logging.

/cc @cramforce @dvoytenko

Comment (dvoytenko) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: This is crazy inconvenient. Ideally we'd find a way to do stripping before renaming/obfuscation. No idea why they thought it'd be a good idea. A stopgap option might be to disable dev/user obfuscation? Obviously that's a bit annoying too, but since we considering to doing them as globals, it doesn't really matter. And most of uses of these APIs are `dev.fine` and `dev.assert` - so it might be ok anyway.

Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: yeah, i might just invoke `addCustomPass` which i think allows me to walk the AST. will try it out

Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: so ive been researching it some more and its not that its running after optimization, its that it runs against the es5 -> es6 transpiled code meaning its walking through something that looks like the screenshot below (where its already been prefixed with the namespace):

screen shot 2016-03-17 at 1 17 30 am

File: third_party/closure-compiler/runner/src/AmpPass.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  }
+  
+  boolean qualifiedNameEndsWithStripType(Node n) {
+    String name = n.getQualifiedName();
+    return qualifiedNameEndsWithStripType(name);
+  }
+
+  boolean qualifiedNameEndsWithStripType(String name) {
+    if (name != null) {
+      return name.endsWith("log.dev.fine");
```
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpPass.java]: @dvoytenko @cramforce this is what i came up with which is based on the closure compilers' `stripTypePrefixes` code, this code is more or less a `stripTypeSuffixes` (which isnt implemented)

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    return options;
+  }
+
+  public static void main(String[] args) {
+    AmpCommandLineRunner runner = new AmpCommandLineRunner(args);
+    if (runner.shouldRunCompiler()) {
+      runner.run();
+    }
+    if (runner.hasErrors()) {
+      System.exit(-1);
```
Comment (cramforce) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: Not `1`? Do they get printed out?

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+  protected AmpCommandLineRunner(String[] args) {
+    super(args);
+  }
+
+  @Override
+  protected CompilerOptions createOptions() {
+    CompilerOptions options = super.createOptions();
+    AmpPass ampPass = new AmpPass(getCompiler(), suffixTypes);
+    options.addCustomPass(CustomPassExecutionTime.BEFORE_OPTIMIZATIONS, ampPass);
```
Comment (cramforce) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: Any reason to not use the builtin strip prefix support?

File: third_party/closure-compiler/runner/src/AmpPass.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import java.util.Set;
+
+import com.google.javascript.jscomp.AbstractCompiler;
+import com.google.javascript.jscomp.HotSwapCompilerPass;
+import com.google.javascript.jscomp.NodeTraversal;
+import com.google.javascript.jscomp.NodeTraversal.AbstractPostOrderCallback;
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+
+class AmpPass extends AbstractPostOrderCallback implements HotSwapCompilerPass {
```
Comment (cramforce) [inline third_party/closure-compiler/runner/src/AmpPass.java]: More comments please (if this class is needed)

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 5 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: @cramforce see my commend here https://github.com/ampproject/amphtml/pull/2593#discussion_r56468684

unfortunately the `stripPrefix` (whether by `type` or `name`) operate on the module prefixed variable (guessing the es6 to es5 transpilation) so for it to strip the prefix type "dev.log" for example you would have to need to know it is actually named "module$__$__$__$extensions$amp_test$0_1$log.dev" during that phase (possible?). `name` would work but it doesn't allow for "GETPROP" (dot operator) so you can only give it "dev" or "log" for example which feels very unsafe.

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 4 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: this is an exact copy of the main code in the super class, and in my testing the errors/warning do get printed out.

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 5 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: @cramforce there is a way to get `stripPrefix` to work and that is if we assign it to the global "AMP" namespace and use it like `AMP.dev.fine("text")` since that is never mangled or renamed it can strip it. this would mean though that the `log` module would need to be the initialized even earlier as it might be needed right away.

Comment (dvoytenko) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: @cramforce @erwinmombay I'm not a huge fun of something like this for main binary - I'd rather avoid an extra module loading dependency. We may decide that we no longer want to package `log.js` into every extension and instead use `AMP.dev` and `AMP.user` - but that should be an independent thing.

File: third_party/closure-compiler/runner/.classpath [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="lib" path="/Users/erwinm/dev/amphtml/third_party/closure-compiler/compiler.jar"/>
```
Comment (dvoytenko) [inline third_party/closure-compiler/runner/.classpath]: Is this file actually needed? If so - can we make this a local file path?

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 5 above)
Comment (dvoytenko) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: @erwinmombay @cramforce Not sure if there's a more direct facility in closure compiler, but generally this rings true to do it `BEFORE_OPTIMIZATIONS` since stripping calls can lead to more things to optimize...

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 4 above)
Comment (dvoytenko) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: This library is on GitHub? Can we sprinkle links/references to the "standard" code here?

File: third_party/closure-compiler/runner/.classpath (hunk 7 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/.classpath]: not needed, will clean up the project artifacts

File: third_party/closure-compiler/runner/src/AmpCommandLineRunner.java (hunk 4 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpCommandLineRunner.java]: yep, will do.

File: third_party/closure-compiler/src/AmpCommandLineRunner.java (hunk 1 above)
Comment (erwinmombay) [inline third_party/closure-compiler/src/AmpCommandLineRunner.java]: added JUnit tests.

File: third_party/closure-compiler/runner/src/AmpPass.java (hunk 6 above)
Comment (erwinmombay) [inline third_party/closure-compiler/runner/src/AmpPass.java]: done.

File: test/size.txt [hunk 8]
```diff
@@ -1,38 +1,38 @@
       max   |         min   |       gzip   |   file
       ---   |         ---   |        ---   |   ---
-611.74 kB   |   178.04 kB   |   49.15 kB   |   v0.js / amp.js
```
Comment (erwinmombay) [inline test/size.txt]: /cc @cramforce not too big of a win once gzip'd but a good start

**Comment by erwinmombay** [comment]
@cramforce this is ready for a full review. PTAL

**Comment by erwinmombay** [comment]
@cramforce one slightly big issue is the addition of the 2 semi large jar files. runner.jar (5.9mb) and compiler-and-tests.jar (10.4mb, and a redundant build of compiler.jar with tests so i can inherit from base test classes for testing)

File: build-system/runner/amp-test/0.1/test.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+function main() {
+  let b = dev.assert('assert');
+  dev.fine('fine');
+  dev.warn('warn');
+  var a = 1 + 1;
+  return a;
+}
+
+function sideEffectAssert() {
```
Comment (cramforce) [inline build-system/runner/amp-test/0.1/test.js]: This is not actually a side effect. Ideally this would compile to
`return 'wahhh'`

File: build-system/runner/src/org/ampproject/AmpPass.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import com.google.javascript.jscomp.NodeTraversal;
+import com.google.javascript.jscomp.NodeTraversal.AbstractPostOrderCallback;
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+
+/**
+ * Does a `stripTypeSuffix` which currently can't be done through
+ * the normal `strip` mechanisms provided by closure compiler.
+ * Some of the known mechanisms we tried before writing our own compiler pass
+ * are setStripTypes, setStripTypePrefixes, setStripNameSuffixes, setStripNamePrefixes.
```
Comment (cramforce) [inline build-system/runner/src/org/ampproject/AmpPass.java]: Great comment.

File: build-system/runner/amp-test/0.1/test.js (hunk 9 above)
Comment (cramforce) [inline build-system/runner/amp-test/0.1/test.js]: If that is not what you currently do it may be fine for now, but we should be correct on terminology.

For side effects I'd expects something like

`dev.assert(document.body.appendChild(…))`

to compile to

`document.body.appendChild(…)`

File: test/size.txt [hunk 11]
```diff
@@ -1,38 +1,38 @@
...
+143.99 kB   |    44.74 kB   |   15.59 kB   |   v0/amp-mustache-0.1.js
+120.43 kB   |    30.61 kB   |    9.39 kB   |   v0/amp-pinterest-0.1.js
+ 70.64 kB   |    10.24 kB   |    3.95 kB   |   v0/amp-reach-player-0.1.js
+  95.3 kB   |    17.16 kB   |     6.1 kB   |   v0/amp-sidebar-0.1.js
+105.86 kB   |    17.94 kB   |    6.33 kB   |   v0/amp-slides-0.1.js
+114.12 kB   |    25.01 kB   |    8.89 kB   |   v0/amp-social-share-0.1.js
+ 71.21 kB   |    10.43 kB   |    4.02 kB   |   v0/amp-soundcloud-0.1.js
+ 79.18 kB   |    12.93 kB   |    4.67 kB   |   v0/amp-springboard-player-0.1.js
+132.13 kB   |    24.38 kB   |    8.74 kB   |   v0/amp-twitter-0.1.js
+278.81 kB   |    62.34 kB   |   18.34 kB   |   v0/amp-user-notification-0.1.js
```
Comment (cramforce) [inline test/size.txt]: Out of curiosity: Why is this so big?

File: build-system/runner/src/org/ampproject/AmpPass.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    if (call == null) {
+      return;
+    }
+    Node expression = call.getFirstChild();
+    if (expression == null) {
+      return;
+    }
+
+    if (qualifiedNameEndsWithStripType(expression)) {
+      if (parent.isExprResult()) {
```
Comment (cramforce) [inline build-system/runner/src/org/ampproject/AmpPass.java]: You currently just kill the full statement, right?

I wonder if, for asserts, you should just alway copy the first argument here. It would be DCEed if it is side effect free.

File: third_party/closure-compiler/Messages.properties [hunk 13]
```diff
@@ -0,0 +1,335 @@
+#
```
Comment (cramforce) [inline third_party/closure-compiler/Messages.properties]: Why do we need this and the next file?

Comment (erwinmombay) [inline third_party/closure-compiler/Messages.properties]: from what i can tell they are not exported in the jar compilation, and they are needed for reporting AST errors during test.

File: build-system/runner/amp-test/0.1/test.js (hunk 9 above)
Comment (erwinmombay) [inline build-system/runner/amp-test/0.1/test.js]: haha sorry, this file is really just for testing. i'll remove this

File: test/size.txt (hunk 11 above)
Comment (erwinmombay) [inline test/size.txt]: thats a good question. it looks like it just has a lot of small dependencies on core that are not abstracted out into services.

notable ones:
- url-replacement (surprised how big this is)
- style
- xhr
- storage
- log (this is common and the easiest one to fix across all extensions)

File: build-system/runner/src/org/ampproject/AmpPass.java (hunk 12 above)
Comment (erwinmombay) [inline build-system/runner/src/org/ampproject/AmpPass.java]: right. should we just go ahead and change that for all typeSuffixes we look for or specifically just for assert?

File: third_party/closure-compiler/Messages.properties (hunk 13 above)
Comment (erwinmombay) [inline third_party/closure-compiler/Messages.properties]: they are also part of the closure-compiler source code (just not bundled). do we need to separate this out into third_party/rhino ?

File: build-system/runner/src/org/ampproject/AmpPass.java (hunk 12 above)
Comment (cramforce) [inline build-system/runner/src/org/ampproject/AmpPass.java]: Just for assert. We should just completely kill logging.

Comment (erwinmombay) [inline build-system/runner/src/org/ampproject/AmpPass.java]: done.

File: build-system/runner/test/org/ampproject/AmpPassTest.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        LINE_JOINER.join(
+             "(function() {",
+             "  var log = { dev: { assert: function() {} } };",
+             "\"hello world\";",
+             "  console.log('this is preserved');",
+            "})()"));
+  }
+
+  public void testDevAssertPreserveFirstArg() throws Exception {
+    test(
```
Comment (erwinmombay) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: /cc @cramforce

**Comment by erwinmombay** [comment]
@cramforce PTAL

File: build-system/runner/test/org/ampproject/AmpPassTest.java (hunk 14 above)
Comment (cramforce) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: Excellent.

**Comment by cramforce** [comment]
LGTM

File: test/size.txt [hunk 15]
```diff
@@ -1,38 +1,38 @@
       max   |         min   |       gzip   |   file
       ---   |         ---   |        ---   |   ---
-612.52 kB   |   178.05 kB   |   49.15 kB   |   v0.js / amp.js
```
Comment (erwinmombay) [inline test/size.txt]: /cc @dvoytenko for log elimination new sizes

Comment (cramforce) [inline test/size.txt]: Nice win.
